library/x509_crt.c

```diff
@@ -83,6 +83,342 @@
+    x509_buf_to_buf_raw( &frame->serial, &crt->serial );
+    x509_buf_to_buf_raw( &frame->pubkey_raw, &crt->pk_raw );
+    x509_buf_to_buf_raw( &frame->issuer_raw, &crt->issuer_raw );
+    x509_buf_to_buf_raw( &frame->subject_raw, &crt->subject_raw );
+    x509_buf_to_buf_raw( &frame->subject_id, &crt->subject_id );
+    x509_buf_to_buf_raw( &frame->issuer_id, &crt->issuer_id );
+    x509_buf_to_buf_raw( &frame->sig, &crt->sig );
+    x509_buf_to_buf_raw( &frame->v3_ext, &crt->v3_ext );
+
+    /* The legacy CRT structure doesn't explicitly contain
+     * the `AlgorithmIdentifier` bounds; however, those can
+     * be inferred from the surrounding (mandatory) `SerialNumber`
+     * and `Issuer` fields. */
+    frame->sig_alg.p = crt->serial.p + crt->serial.len;
+    frame->sig_alg.len = crt->issuer_raw.p - frame->sig_alg.p;
+
+    return( x509_crt_frame_parse_ext( frame ) );
+#endif /* !MBEDTLS_X509_ON_DEMAND_PARSING */
+}
+
+int mbedtls_x509_crt_cache_provide_pk( mbedtls_x509_crt const *crt )
+{
+    mbedtls_x509_crt_cache *cache = crt->cache;
+    mbedtls_pk_context *pk;
+
+    if( cache->pk != NULL )
+        return( 0 );
+
+    pk = mbedtls_calloc( 1, sizeof( mbedtls_pk_context ) );
+    if( pk == NULL )
+        return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+    cache->pk = pk;
+
+#if !defined(MBEDTLS_X509_ON_DEMAND_PARSING)
+    *pk = crt->pk;
+    return( 0 );
+#else
+    {
+        mbedtls_x509_buf_raw pk_raw = cache->pk_raw;
+        return( mbedtls_pk_parse_subpubkey( &pk_raw.p,
+                                            pk_raw.p + pk_raw.len,
+                                            pk ) );
+    }
+#endif /* MBEDTLS_X509_ON_DEMAND_PARSING */
+}
+
+static void x509_crt_cache_init( mbedtls_x509_crt_cache *cache )
+{
+    memset( cache, 0, sizeof( *cache ) );
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_init( &cache->frame_mutex );
+    mbedtls_mutex_init( &cache->pk_mutex );
+#endif
+}
+
+static void x509_crt_cache_clear_pk( mbedtls_x509_crt_cache *cache )
+{
+#if !defined(MBEDTLS_X509_ON_DEMAND_PARSING)
+    /* The cache holds a shallow copy of the PK context
+     * in the legacy struct, so don't free PK context. */
+    mbedtls_free( cache->pk );
+#else
+    mbedtls_pk_free( cache->pk );
+    mbedtls_free( cache->pk );
+#endif /* MBEDTLS_X509_ON_DEMAND_PARSING */
+
+    cache->pk = NULL;
+}
+
+static void x509_crt_cache_clear_frame( mbedtls_x509_crt_cache *cache )
+{
+    mbedtls_free( cache->frame );
+    cache->frame = NULL;
+}
+
+static void x509_crt_cache_free( mbedtls_x509_crt_cache *cache )
+{
+    if( cache == NULL )
+        return;
+
+#if defined(MBEDTLS_THREADING_C)
+    mbedtls_mutex_free( &cache->frame_mutex );
+    mbedtls_mutex_free( &cache->pk_mutex );
+#endif
+
+    x509_crt_cache_clear_frame( cache );
+    x509_crt_cache_clear_pk( cache );
+
+    memset( cache, 0, sizeof( *cache ) );
+}
+
+int mbedtls_x509_crt_get_subject_alt_names( mbedtls_x509_crt const *crt,
```

Comment:
@RonEld So far, I see the following objects in our discussion:
1. Allowing users an easy switch to the new API when this PR is merged.
2. Giving users convenience in accessing various SAN types once we start to support more than `dNSName`.
3. Modularity - we don't want to increase code-size with features that many of our users won't need.

My function prototype mainly aims to address point 1., while you are more interested in 2. (I think).

I am reluctant to go with just 2. because it isn't aligned with 1., and it also adds complexity not everyone might need, hence not aligning well to 3.

What do you think about the following approach: 
* We keep the legacy `mbedtls_x509_sequence` structure as the first 'layer' of accessing the SAN extension, and we leave `mbedtls_x509_crt_get_subject_alt_names()` as-is to allow an easy switch for users that previously used `mbedtls_x509_crt::subject_alt_names` which now no longer exists. That solves 1. 
* Next, for 2., we add your parser as a 'second layer' function which takes an ASN.1 buffer holding a SAN entry and returns a parsed representation of the SAN entry. I am not sure if that should exactly look like your `mbedtls_x509_subject_alternative_name`, which in itself is a linked list, duplicating the original `subject_alt_names` linked list (which makes sense under your assumption that it _replaces_ the former), but that's rather a detail. 
* With this separation, 3. would be sufficiently considered because we have rather small functions serving single purposes, and link time optimization will drop them unless the user needs them. 

What do you think?

Reply:
yes, I am more interested in 2. WE have got several request to support IP addresses as SAN, so I believe it is close to be inserted.
In addition, supporting `otherName` already exists in #2530, so it's very soon that this needs to change.
Since we are already adding \ changing API, I think it's reasonable to change  the protptype os the new API `mbedtls_x509_crt_get_subject_alt_names()` , before it is being introduced, rather than introducing yet another new API that will convert the `mbedtls_x509_sequence` to a different structure.
I have already changed in #2530 to return the sequence of `mbedtls_x509_sequence` items. And then, if user chooses, call a public API to `convert` it to `mbedtls_x509_subject_alternative_name`. I think this is similar to the spirit of this API, except the conversion should happen from the raw buffer to `mbedtls_x509_subject_alternative_name`.

Reply:
Yes, I see your point.

What I would like to avoid is the addition of parsing logic for more complex SAN types that will be included in the library even if application code doesn't need it. The above proposal would address this by having link time GC remove the respective code when it's not being used. With your proposal, we would probably need a new compile-time option.

But on the other hand, if we start supporting more complex SAN types, we should at least check structural integrity during CRT parsing, so the bulk of the parsing functions would be in the core CRT code anyway, and the only way to allow getting rid of them would be through a compile-time option. 

> ... before it is being introduced, rather than introducing yet another new API that will convert the mbedtls_x509_sequence to a different structure.

To make sure that we're thinking of the same thing, let me elaborate on my proposal, so you can point me more precisely at where you spot problems.

In #2530, you propose the following structure:
```c
/*
 * A structure for holding the parsed Subject Alternative Name, according to type
 */
typedef struct mbedtls_x509_subject_alternative_name
{
    int type;                              /**< The SAN type, value of MBEDTLS_X509_SAN_XXX. */
    union {
        mbedtls_x509_san_other_name other_name; /**< The otherName supported type. */
        mbedtls_x509_buf   unstructured_name; /**< The buffer for the un constructed types. Only dnsName currently supported */
    }
    san; /**< A union of the supported SAN types */
    struct mbedtls_x509_subject_alternative_name *next; /**< The next SAN in the list. */
}
mbedtls_x509_subject_alternative_name;
```
My proposal would be to remove the linked list component of this and making it standalone and not referencing any dynamic memory, for example:
```c
typedef struct mbedtls_x509_subject_alternative_name
{
    int type;                              /**< The SAN type, value of MBEDTLS_X509_SAN_XXX. */
    union {
        mbedtls_x509_san_other_name other_name; /**< The otherName supported type. */
        mbedtls_x509_buf   unstructured_name; /**< The buffer for the un constructed types. Only dnsName currently supported */
    }
    parsed; /**< A union of the supported SAN types */
}
mbedtls_x509_subject_alternative_name;
```
Then, we would have a function like
```c
int x509_parse_subject_alternative_name( int tag, 
                                         unsigned char *p, 
                                         unsigned char *end,
                                         mbedtls_x509_subject_alternative_name *dst );
```
or
```c
int x509_parse_subject_alternative_name( mbedtls_x509_buf *raw,
                                         mbedtls_x509_subject_alternative_name *dst );
```
which takes an ASN.1 buffer holding a SAN component, and fills the target structure with the parsed representation. 

This would then we used in two places: (1) We use it ourselves during CRT parsing, passing it as the callback to the `mbedtls_asn1_traverse_sequence_of()` to check for structural integrity (but always using the same `dst` because we don't actually want to save all parsed presentation of the SAN components). (2) The user can call it on the entries of the `mbedtls_x509_crt::subject_alt_names` linked list, if needed.

This would align the introduction of `mbedtls_x509_subject_alternative_name` to the goal of this PR that you shouldn't be forced to use dynamic memory if you want to parse a CRT, and indeed users wouldn't need to: Like us during CRT parsing, they could use `x509_parse_subject_alternative_name()` in conjunction with `mbedtls_asn1_traverse_sequence_of()` to parse complex SAN components without dynamic memory.

Reply:
maybe I am missing something, but even in this PR, when user choose to parse the SAN, it return a sequence of the dnsName items, not a single item.

My concern is in the public function prototype:
~~~
int mbedtls_x509_crt_get_subject_alt_names( mbedtls_x509_crt const *crt,
                                            mbedtls_x509_sequence **subj_alt )
~~~
This extracts from the certificate a sequence of SAN. The implementation is currently not important, we can leave it `dnsName` for this PR. But, if we expose a public API to return only non constructed types, it will be problematic to add support for complex SAN types.
If remove the linked list component will keep same behavior, I don't mind, but I don't see how it can happen.
It's not mandatory that there are single entries of complex components.

Reply:
> But, if we expose a public API to return only non constructed types, it will be problematic to add support for complex SAN types.

Where is the problem with the approach I described?

Reply:
I don't want to add another API in the future, if we can avoid.
`mbedtls_x509_crt_get_subject_alt_names` is an API that will return a sequence of `mbedtls_asn1_buf` . This will hold the raw data of the SAN. For complex SAN types, it will require an additional call by the user to parse these. I prefer it to be parsed in this function, to avoid unneccessary calls by the user, as it actually requires the user to do some X509 related functionality in order to extract the exact complex struct.

Reply:
@RonEld For complex SANs to come, we will need some public API that allows users to parse them from raw data without the allocation of dynamically linked lists, in the spirit of the rest of this PR. Hence, if we want to maintain _some_ form of dynamic linked list for convenience of users, it is unavoidable to have two new APIs. The question, then, is: Do we want to have overlap between the two APIs or not? With your proposal to do the SAN component parsing as part of the API that returns the linked list representation, we have the SAN component parsing in both APIs. In contrast, if we keep the API as proposed in this PR, it is separate, and - to me - feels more modular. Also, it is closer to the previous `subject_alt_names` field in `mbedtls_x509_crt`, and hence will require less work to migrate. That's why I prefer the current approach.

Reply:
But this PR also allocated a dynamically linked list, in the public API. See `x509_get_subject_alt_name_cb`.
I don't see a reason why not have a single dynamic list , of the SAN  struct I suggested, and do it once in the public API.
> Also, it is closer to the previous subject_alt_names field in mbedtls_x509_crt, and hence will require less work to migrate. That's why I prefer the current approach.

We currently don't have a public API to get the SAN, so I don't think this is relevant. As we are introducing new APi, we can define it is we see fit.

Reply:
> We currently don't have a public API to get the SAN, so I don't think this is relevant. 

We have a field which contains the linked list presentation of raw SANs, and that's what I'm comparing new APIs against.

> I don't see a reason why not have a single dynamic list , of the SAN struct I suggested, and do it once in the public API.

Please see my last comment: I also want a single linked list, but one that contains raw SAN data, plus another public API for parsing SAN components from raw data, _which we will need anyway_. If we expose (1) a public API returning a linked list of parsed SAN components, and (2) a public API for parsing raw SAN components, there is an internal dependency and overlap between both, which I want to avoid.

Reply:
See what I mean, in [2530](https://github.com/ARMmbed/mbedtls/pull/2530/files#diff-7eb92ee9f705f88338240a484631c19cR1773)
This is a single public function that parses the raw data of SAN.
Internally, it calls two different helper functions, according to the type.
I would like to avoid having the user having to check what SAN type it is, as it requires some knowledge that I don't think the user should necessary have. 
The  single linked-list-generating API, but one that contains raw SAN data should exist, but I don't think that it is necessary to be a public API, as the user is eventually interested in the SAN itself

Reply:
> This is a single public function that parses the raw data of SAN. Internally, it calls two different helper functions, according to the type.

Yes, that's what I want, too. Good, we've agreed we'll have this new raw SAN parsing API. But this cannot use the type `mbedtls_x509_subject_alternative_name` as you've proposed it, because that's a linked list entry.

> I would like to avoid having the user having to check what SAN type it is, as it requires some knowledge that I don't think the user should necessary have.

I have no strong feelings on that one, I'm OK either way.

> The single linked-list-generating API, but one that contains raw SAN data should exist, but I don't think that it is necessary to be a public API, as the user is eventually interested in the SAN itself

Are you suggesting to not have any linked-list generating API? IMHO that's not acceptable because we must provide a replacement for the linked lists currently contained in `mbedtls_x509_crt` (and there won't be a linked list presentation of parsed SAN components in `mbedtls_x509_crt` either, eventually).

Ron, please re-consider https://github.com/ARMmbed/mbedtls/pull/2478#discussion_r280412095, I think we're really talking about the same thing here. I can't reconcile your disagreement on that proposal with your latest comments.

Reply:
>But this cannot use the type mbedtls_x509_subject_alternative_name as you've proposed it, because that's a linked list entry.

Yes, the idea is to return a linked list.

Perhaps I didn't understand your model. Are you suggesting the two public APIs:
1. Extract the SAN into a sequence of raw data.
2. For each item, parse it according to type.

Reply:
@RonEld Yes, exactly. (1) mainly for convenience and backwards compatibility for users, and (2) to be used either on top of (1) or on top of a raw ASN.1 traversal.